### PR TITLE
[Ops] Feat: Add triton-ascend backend support for FLA module kernels

### PR DESCRIPTION
**Pull request by zheliuyu**
## Summary

This PR introduces `triton-ascend` support for a set of commonly used FLA module operators. The design goal is to keep upstream GPU code paths intact while routing NPU execution through a dedicated backend layer at runtime.

When running on NPU (`IS_NPU=True`), decorated entry points automatically dispatch to implementations under `fla/modules/backends/triton_ascend/`. On GPU/CPU, behavior remains unchanged.

We only adapted kernels in `modules/` in this first PR since they're relatively basic. The rest will follow the same pattern shown here — stay tuned!

- [x] fla/modules/
- [ ] fla/ops/
- [ ] fla/layers/
- [ ] fla/models/

---

## Motivation

Flash Linear Attention (FLA) module kernels were originally written and tuned for GPU Triton (CUDA/HIP). Running them on NPU via `triton-ascend` requires targeted adaptations due to:

- Different compiler behavior (MLIR passes, mask handling, grid limits)
- Stricter on-chip memory (UB) constraints
- Reduced autotuning flexibility compared to GPU Triton
- Incompatibilities with certain GPU-oriented patterns (e.g. `make_block_ptr`, large warp counts, specific block-tile shapes)

Rather than branching heavily inside upstream files, this PR isolates NPU logic in a backend directory and uses a small dispatch hook at the public API layer.

---

### Architecture

```
User / model code (unchanged)
        │
        ▼
fla/modules/<module>.py              # Public API + @dispatch('modules') hooks only
        │
        ▼
fla/ops/backends                     # Runtime backend selection (BackendRegistry)
        │
        ▼
fla/modules/backends/triton_ascend/  #  NPU Triton kernels + launchers
        ├── __init__.py              # TritonAscendBackend registration
        ├── rotary.py
        ├── fused_cross_entropy.py
        ├── fused_linear_cross_entropy.py
        ├── activations.py
        ├── fused_kl_div.py
        ├── layernorm.py
        ├── grpo.py
        └── causal_conv1d.py
```

**Note: The remaining files under `modules/` will be reused.**

### Principles

1. **Minimal upstream diff** — upstream files only add `@dispatch('modules')` (and small glue where required, e.g. conv layout detection). No NPU kernel logic in GPU sources.
2. **Automatic selection** — `TritonAscendBackend.is_available()` returns `True` when `fla.utils.IS_NPU` is set.
3. **Opt-out** — set `FLA_DISABLE_BACKEND_DISPATCH=1` to bypass dispatch and use default implementations.

### Registry

- `fla/ops/backends/__init__.py` maps `'modules'` → `'fla.modules.backends'`.
- `fla/modules/backends/__init__.py` registers `TritonAscendBackend` via `BackendRegistry("modules")`.

---

## Modules adapted

| Module | Upstream hooks | NPU file |
|--------|----------------|----------|
| Rotary embedding | `fla/modules/rotary.py` | `rotary.py` |
| Fused cross-entropy | `fla/modules/fused_cross_entropy.py` | `fused_cross_entropy.py` |
| Fused linear cross-entropy | `fla/modules/fused_linear_cross_entropy.py` | `fused_linear_cross_entropy.py` |
| Activations (sigmoid, logsigmoid, swish, swiglu) | `fla/modules/activations.py` | `activations.py` |
| Fused KL divergence | `fla/modules/fused_kl_div.py` | `fused_kl_div.py` |
| LayerNorm | `fla/modules/layernorm.py` | `layernorm.py` |
| GRPO loss | `fla/modules/grpo.py` | `grpo.py` |
| Causal Conv1d | `fla/modules/conv/triton/ops.py` | `causal_conv1d.py` |

---

### Shared across modules

- `STATIC_WARPS = 2` on NPU (vs 16–32 on GPU) for compiler stability
- Grid clamping to stay under NPU’s 65,535 block limit
- Stride-aware Triton loads for non-contiguous tensors (activations, conv backward `dy`, etc.)

### Causal Conv1d

- Separate fwd/bwd/update/states kernels using explicit pointer loads (not GPU `make_block_ptr`)
- Swish/silu activation and residual via standalone Triton kernels (`_silu_kernel`, `_silu_bwd_kernel`, `_add_kernel`) outside the conv kernel
- `_npu_tile_config` / `_npu_bwd_tile_config` for UB-safe `BD` / `BT`
- Recompute `chunk_indices` whenever `BT` changes (varlen)
- `causal_conv1d_bwd_seq_kernel` for bf16 short sequences (`T ≤ 16`) to avoid triton-ascend MLIR mask bugs
- `_clamp_bd_for_grid` for large `D` (e.g. 8192)

### Fused KL Div

- `_npu_vocab_block_size()` for large vocabularies
- Contiguous logits before `torch.mm`
- Backward block sizing to avoid grid overflow

### Cross-entropy / linear-CE / LayerNorm / GRPO / Activations

- NPU tile sizes, warps, and launch grids tuned per operator
- Non-contiguous layout support where tests require it (e.g. QKV-style views)

---

## Test plan

Verified on NPU(Atlas 800T A3) — all relevant `tests/modules/` cases pass:

截屏2026-06-02 10 58 08

### Commands

```bash
# Full module test suite on NPU
pytest tests/modules/
```

---

## Checklist

- [x] NPU logic isolated under `fla/modules/backends/triton_ascend/`
- [x] Upstream files contain only minimal `@dispatch` hooks
- [x] `tests/modules/` passing on Ascend
- [x] GPU sources preserved for upstream merge compatibility

### CODE REVIEW

**Comment by zhiyuan1i** [comment]
I’m very grateful for such high-quality MRs. Since FLA lacks the necessary infrastructure for NPU support, I’m curious whether it’s possible to use preemptive NPU resources to ensure the accuracy of subsequent MRs.

**Comment by zheliuyu** [comment]
@zhiyuan1i Thanks for your feedback — you raised a really important point. I noticed a similar concern you mentioned in https://github.com/fla-org/flash-linear-attention/pull/737#issuecomment-3869679347 as well. I've already requested NPU machine resources from the infra team, and they should get back to us within about a week. 

Additionally,  I'd like to get your take on the NPU CI workflow code — do you think it's better to include it in this PR, or submit it as a separate follow-up PR?

**Comment by zhiyuan1i** [comment]
> @zhiyuan1i Thanks for your feedback — you raised a really important point. I noticed a similar concern you mentioned in [#737 (comment)](https://github.com/fla-org/flash-linear-attention/pull/737#issuecomment-3869679347) as well. I've already requested NPU machine resources from the infra team, and they should get back to us within about a week.
> 
> Additionally, I'd like to get your take on the NPU CI workflow code — do you think it's better to include it in this PR, or submit it as a separate follow-up PR?

Maybe in a later PR, separate or refactor the CI to ensure GPU tests only run on non-NPU backends by default. You could also trigger a separate workflow to test only NPU, and generate a compatibility report.

**Comment by zheliuyu** [comment]
> Maybe in a later PR, separate or refactor the CI to ensure GPU tests only run on non-NPU backends by default. You could also trigger a separate workflow to test only NPU, and generate a compatibility report.

We plan to go with the second option and set up a dedicated NPU workflow, referencing:
- https://github.com/ascend-gha-runners/docs/blob/main/docs/user-manual-gha-en.md
- https://github.com/ascend-gha-runners/docs/blob/main/docs/user-manual-gha-zh.md

Additionally, we may need to coordinate with you on runner access and permissions down the line. 🤝
